core/src/main/java/com/yahoo/oak/InternalOakHash.java

```diff
@@ -0,0 +1,495 @@
+/*
+ * Copyright 2020, Verizon Media.
+ * Licensed under the terms of the Apache 2.0 license.
+ * Please see LICENSE file in the project root for terms.
+ */
+
+package com.yahoo.oak;
+
+import java.nio.ByteBuffer;
+import java.util.function.Consumer;
+
+class InternalOakHash<K, V> extends InternalOakBasics<K, V> {
+    /*-------------- Members --------------*/
+    private final FirstLevelHashArray<K, V> hashArray;    // first level of indexing
+    private final OakComparator<K> comparator;
+    private final OakSerializer<K> keySerializer;
+    private final OakSerializer<V> valueSerializer;
+    private final ValueUtils valueOperator;
+
+    private static final int DEFAULT_MOST_SIGN_BITS_NUM = 16;
+    static final int USE_DEFAULT_FIRST_TO_SECOND_BITS_PARTITION = -1;
+
+    /*-------------- Constructors --------------*/
+    InternalOakHash(OakSerializer<K> keySerializer, OakSerializer<V> valueSerializer,
+        OakComparator<K> oakComparator, MemoryManager vMM, MemoryManager kMM,
+        ValueUtils valueOperator, int firstLevelBitSize, int secondLevelBitSize) {
+
+        super(vMM, kMM);
+        this.keySerializer = keySerializer;
+        this.valueSerializer = valueSerializer;
+        this.comparator = oakComparator;
+        this.valueOperator = valueOperator;
+
+        int msbForFirstLevelHash =
+            (firstLevelBitSize == USE_DEFAULT_FIRST_TO_SECOND_BITS_PARTITION)
+                ? DEFAULT_MOST_SIGN_BITS_NUM : firstLevelBitSize;
+
+        this.hashArray =
+            new FirstLevelHashArray<K, V>(msbForFirstLevelHash, secondLevelBitSize,
+                this.size, vMM, kMM, oakComparator,
+                keySerializer, valueSerializer, 1);
+
+    }
+
+    /*-------------- Generic to specific rebalance --------------*/
+    @Override
+    protected void rebalanceBasic(BasicChunk<K, V> basicChunk) {
+        rebalance((HashChunk<K, V>) basicChunk); // exception will be triggered on wrong type
+    }
+
+    /**
+     * @param c - OrderedChunk to rebalance
+     */
+    private void rebalance(HashChunk<K, V> c) {
+        // temporary throw an exception in order not to miss this erroneous behavior while debug
+        throw new RuntimeException("Not yet implemented rebalance of a HashChunk was requested");
```

Comment:
You can use `UnsupportedOperationException` in this case.

Reply:
changed